File: SPCode/Modules/PopulateSupport.vb

Module PopulateSupport



    Public Function BuildSetupList(ByRef TreeViewControl As Object, ByRef ImageListControl As ImageList, ByRef SourceData As DataTable, ByVal ParentField As String, _
ByVal CompareField As String, ByVal DisplayField As String, ByVal NodeIDField As String, Optional ByVal ClearTree As Boolean = True) As Boolean
        Cursor.Current = Cursors.WaitCursor

        Dim tvwTreeView As TreeView = CType(TreeViewControl, TreeView)
        Dim imlImageList As ImageList = ImageListControl

        tvwTreeView.BeginUpdate()
        If ClearTree Then
            tvwTreeView.Nodes.Clear()
        End If
        tvwTreeView.ImageList = imlImageList

        Dim dvTreeTable As New DataView, currentNode As TreeNode

        dvTreeTable.Table = SourceData



        If ClearTree = False Then 'Might be a root node
            tvwTreeView.BeginUpdate()
            Try
                currentNode = tvwTreeView.Nodes(0)
                Call BuildSetupNode(currentNode, dvTreeTable, _
                Nothing, ParentField, CompareField, NodeIDField, DisplayField)

                currentNode.Expand()
            Catch ex As Exception

            End Try
            tvwTreeView.EndUpdate()
        Else
            dvTreeTable.RowFilter = " IsNull(" & ParentField & ",'') = '' "
            For Each drvRow As DataRowView In dvTreeTable
                tvwTreeView.BeginUpdate()
                currentNode = tvwTreeView.Nodes.Add(drvRow( _
                DisplayField).ToString)
                currentNode.Tag = drvRow(NodeIDField).ToString
                Call BuildNode(currentNode, dvTreeTable, drvRow, _
                ParentField, CompareField, NodeIDField, DisplayField)
                currentNode.Expand()
                tvwTreeView.EndUpdate()
            Next

        End If

        tvwTreeView.EndUpdate()
        Cursor.Current = Cursors.Default
        Return True
    End Function
    Public Function BuildTree(ByRef TreeViewControl As TreeView, ByRef ImageListControl As ImageList, ByRef SourceData As DataTable, ByVal ParentField As String, _
    ByVal CompareField As String, ByVal DisplayField As String, ByVal NodeIDField As String, Optional ByVal ClearTree As Boolean = True) As Boolean
        Cursor.Current = Cursors.WaitCursor

        Dim tvwTreeView As TreeView = CType(TreeViewControl, TreeView)
        Dim imlImageList As ImageList = ImageListControl

        tvwTreeView.BeginUpdate()
        If ClearTree Then
            tvwTreeView.Nodes.Clear()
        End If
        tvwTreeView.ImageList = imlImageList


        Dim dvTreeTable As New DataView, currentNode As TreeNode

        dvTreeTable.Table = SourceData



        If ClearTree = False Then 'Might be a root node
            tvwTreeView.BeginUpdate()
            Try
                currentNode = tvwTreeView.Nodes(0)
                Call BuildNode(currentNode, dvTreeTable, _
                Nothing, ParentField, CompareField, NodeIDField, DisplayField)

                currentNode.Expand()
            Catch ex As Exception

            End Try
            tvwTreeView.EndUpdate()
        Else
            dvTreeTable.RowFilter = " IsNull(" & ParentField & ",'') = '' "
            For Each drvRow As DataRowView In dvTreeTable
                tvwTreeView.BeginUpdate()
                currentNode = tvwTreeView.Nodes.Add(drvRow( _
                DisplayField).ToString)
                currentNode.Tag = drvRow(NodeIDField).ToString
                Call BuildNode(currentNode, dvTreeTable, drvRow, _
                ParentField, CompareField, NodeIDField, DisplayField)
                currentNode.Expand()
                tvwTreeView.EndUpdate()
            Next

        End If

        tvwTreeView.EndUpdate()
        Cursor.Current = Cursors.Default
        Return True
    End Function


    Private Sub BuildSetupNode(ByRef currentNode As TreeNode, ByRef sourceData As DataView, ByRef currentRecord As DataRowView, ByVal ParentField As String, ByVal CompareField As String, ByVal NodeIDField As String, ByVal DisplayField As String)
        Dim newData As New DataView(sourceData.Table)
        Dim newNode As TreeNode

        sourceData.RowFilter = vbNullString
        If Not currentRecord Is Nothing Then
            sourceData.RowFilter = ParentField & " = '" & _
            currentRecord(CompareField).ToString & "'"
        End If

        currentNode.TreeView.BeginUpdate()

        For Each drvRow As DataRowView In sourceData
            newData.RowFilter = ParentField & "  = '" & drvRow(CompareField).ToString & "'"
            newNode = currentNode.Nodes.Add(drvRow(DisplayField).ToString)
            newNode.Tag = drvRow(NodeIDField).ToString
            If newData.Count > 0 Then
                Call BuildSetupNode(newNode, newData, drvRow, ParentField, _
                CompareField, NodeIDField, DisplayField)
            End If
        Next

        currentNode.TreeView.EndUpdate()
    End Sub
    Private Sub BuildNode(ByRef currentNode As TreeNode, ByRef sourceData As DataView, ByRef currentRecord As DataRowView, ByVal ParentField As String, ByVal CompareField As String, ByVal NodeIDField As String, ByVal DisplayField As String)
        Dim newData As New DataView(sourceData.Table)
        Dim newNode As TreeNode

        sourceData.RowFilter = vbNullString
        If Not currentRecord Is Nothing Then
            sourceData.RowFilter = ParentField & " = '" & _
            currentRecord(CompareField).ToString & "'"
        End If

        currentNode.TreeView.BeginUpdate()

        For Each drvRow As DataRowView In sourceData
            newData.RowFilter = ParentField & "  = '" & drvRow(CompareField).ToString & "'"
            newNode = currentNode.Nodes.Add(drvRow(DisplayField).ToString)
            newNode.Tag = drvRow(NodeIDField).ToString
            If newData.Count > 0 Then
                Call BuildNode(newNode, newData, drvRow, ParentField, _
                CompareField, NodeIDField, DisplayField)
            End If
        Next

        currentNode.TreeView.EndUpdate()
    End Sub

    Public Function TreeNodeID(ByVal SourceNode As TreeNode, ByVal SelectedIndex As Integer) As Integer
        If SourceNode.Index <> SelectedIndex Then
            For Each curNode As TreeNode In SourceNode.Nodes
                If curNode.Index <> SelectedIndex Then
                    If Not curNode.Nodes Is Nothing Then
                        Call TreeNodeID(curNode, SelectedIndex)
                    End If
                End If
            Next
        Else
            Return CType(SourceNode.Tag, Integer)
        End If
    End Function
    Public Function TreeNodeSearch(ByVal SourceNode As TreeNode, ByVal SearchID As Integer) As TreeNode
        If CType(SourceNode.Tag, Integer) <> SearchID Then
            For Each curNode As TreeNode In SourceNode.Nodes
                If CType(curNode.Tag, Integer) <> SearchID Then
                    If Not curNode.Nodes Is Nothing Then
                        Call TreeNodeSearch(curNode, SearchID)
                    End If
                End If
            Next
        End If
        Return SourceNode
    End Function

End Module

File: SPCode/CS/Modules/PopulateSupport.cs
using System.Text.RegularExpressions;
using System.Diagnostics;
using System;
using System.Drawing.Drawing2D;
using System.Windows.Forms;
using System.Collections;
using System.Drawing;
using Microsoft.VisualBasic;
using System.Data.SqlClient;
using System.Data;
using System.Collections.Generic;
using WeifenLuo.WinFormsUI;
using Microsoft.Win32;
using WeifenLuo;


namespace SoftLogik.Win
{
	sealed class PopulateSupport
	{
		
		
		
		public static bool BuildSetupList(object TreeViewControl, ImageList ImageListControl, DataTable SourceData, string ParentField, string CompareField, string DisplayField, string NodeIDField, bool ClearTree)
		{
			Cursor.Current = Cursors.WaitCursor;
			
			TreeView tvwTreeView = (TreeView) TreeViewControl;
			ImageList imlImageList = ImageListControl;
			
			tvwTreeView.BeginUpdate();
			if (ClearTree)
			{
				tvwTreeView.Nodes.Clear();
			}
			tvwTreeView.ImageList = imlImageList;
			
			DataView dvTreeTable = new DataView();
			TreeNode currentNode;
			
			dvTreeTable.Table = SourceData;
			
			
			
			if (ClearTree == false) //Might be a root node
			{
				tvwTreeView.BeginUpdate();
				try
				{
					currentNode = tvwTreeView.Nodes[0];
					BuildSetupNode(currentNode, dvTreeTable, null, ParentField, CompareField, NodeIDField, DisplayField);
					
					currentNode.Expand();
				}
				catch (Exception)
				{
					
				}
				tvwTreeView.EndUpdate();
			}
			else
			{
				dvTreeTable.RowFilter = " IsNull(" + ParentField + ",\'\') = \'\' ";
				foreach (DataRowView drvRow in dvTreeTable)
				{
					tvwTreeView.BeginUpdate();
					currentNode = tvwTreeView.Nodes.Add(drvRow[DisplayField].ToString());
					currentNode.Tag = drvRow[NodeIDField].ToString();
					BuildNode(currentNode, dvTreeTable, drvRow, ParentField, CompareField, NodeIDField, DisplayField);
					currentNode.Expand();
					tvwTreeView.EndUpdate();
				}
				
			}
			
			tvwTreeView.EndUpdate();
			Cursor.Current = Cursors.Default;
			return true;
		}
		public static bool BuildTree(TreeView TreeViewControl, ImageList ImageListControl, DataTable SourceData, string ParentField, string CompareField, string DisplayField, string NodeIDField, bool ClearTree)
		{
			Cursor.Current = Cursors.WaitCursor;
			
			TreeView tvwTreeView = TreeViewControl;
			ImageList imlImageList = ImageListControl;
			
			tvwTreeView.BeginUpdate();
			if (ClearTree)
			{
				tvwTreeView.Nodes.Clear();
			}
			tvwTreeView.ImageList = imlImageList;
			
			
			DataView dvTreeTable = new DataView();
			TreeNode currentNode;
			
			dvTreeTable.Table = SourceData;
			
			
			
			if (ClearTree == false) //Might be a root node
			{
				tvwTreeView.BeginUpdate();
				try
				{
					currentNode = tvwTreeView.Nodes[0];
					BuildNode(currentNode, dvTreeTable, null, ParentField, CompareField, NodeIDField, DisplayField);
					
					currentNode.Expand();
				}
				catch (Exception)
				{
					
				}
				tvwTreeView.EndUpdate();
			}
			else
			{
				dvTreeTable.RowFilter = " IsNull(" + ParentField + ",\'\') = \'\' ";
				foreach (DataRowView drvRow in dvTreeTable)
				{
					tvwTreeView.BeginUpdate();
					currentNode = tvwTreeView.Nodes.Add(drvRow[DisplayField].ToString());
					currentNode.Tag = drvRow[NodeIDField].ToString();
					BuildNode(currentNode, dvTreeTable, drvRow, ParentField, CompareField, NodeIDField, DisplayField);
					currentNode.Expand();
					tvwTreeView.EndUpdate();
				}
				
			}
			
			tvwTreeView.EndUpdate();
			Cursor.Current = Cursors.Default;
			return true;
		}
		
		
		private static void BuildSetupNode(TreeNode currentNode, DataView sourceData, DataRowView currentRecord, string ParentField, string CompareField, string NodeIDField, string DisplayField)
		{
			DataView newData = new DataView(sourceData.Table);
			TreeNode newNode;
			
			sourceData.RowFilter = Constants.vbNullString;
			if (currentRecord != null)
			{
				sourceData.RowFilter = ParentField + " = \'" + currentRecord[CompareField].ToString() + "\'";
			}
			
			currentNode.TreeView.BeginUpdate();
			
			foreach (DataRowView drvRow in sourceData)
			{
				newData.RowFilter = ParentField + "  = \'" + drvRow[CompareField].ToString() + "\'";
				newNode = currentNode.Nodes.Add(drvRow[DisplayField].ToString());
				newNode.Tag = drvRow[NodeIDField].ToString();
				if (newData.Count > 0)
				{
					BuildSetupNode(newNode, newData, drvRow, ParentField, CompareField, NodeIDField, DisplayField);
				}
			}
			
			currentNode.TreeView.EndUpdate();
		}
		private static void BuildNode(TreeNode currentNode, DataView sourceData, DataRowView currentRecord, string ParentField, string CompareField, string NodeIDField, string DisplayField)
		{
			DataView newData = new DataView(sourceData.Table);
			TreeNode newNode;
			
			sourceData.RowFilter = Constants.vbNullString;
			if (currentRecord != null)
			{
				sourceData.RowFilter = ParentField + " = \'" + currentRecord[CompareField].ToString() + "\'";
			}
			
			currentNode.TreeView.BeginUpdate();
			
			foreach (DataRowView drvRow in sourceData)
			{
				newData.RowFilter = ParentField + "  = \'" + drvRow[CompareField].ToString() + "\'";
				newNode = currentNode.Nodes.Add(drvRow[DisplayField].ToString());
				newNode.Tag = drvRow[NodeIDField].ToString();
				if (newData.Count > 0)
				{
					BuildNode(newNode, newData, drvRow, ParentField, CompareField, NodeIDField, DisplayField);
				}
			}
			
			currentNode.TreeView.EndUpdate();
		}
		
		public static int TreeNodeID(TreeNode SourceNode, int SelectedIndex)
		{
			if (SourceNode.Index != SelectedIndex)
			{
				foreach (TreeNode curNode in SourceNode.Nodes)
				{
					if (curNode.Index != SelectedIndex)
					{
						if (curNode.Nodes != null)
						{
							TreeNodeID(curNode, SelectedIndex);
						}
					}
				}
			}
			else
			{
				return System.Convert.ToInt32(SourceNode.Tag);
			}
		}
		public static TreeNode TreeNodeSearch(TreeNode SourceNode, int SearchID)
		{
			if (System.Convert.ToInt32(SourceNode.Tag) != SearchID)
			{
				foreach (TreeNode curNode in SourceNode.Nodes)
				{
					if (System.Convert.ToInt32(curNode.Tag) != SearchID)
					{
						if (curNode.Nodes != null)
						{
							TreeNodeSearch(curNode, SearchID);
						}
					}
				}
			}
			return SourceNode;
		}
		
	}
	
	
}
